created 2 classes (Book and LibrarySystem)

FilippovichSergey opened the pull request:
solved the task Библиотека Вестеросса

FilippovichSergey committed:
created 2 classes (Book and LibrarySystem)

MarselYsup commented on src/main/java/school/faang/Book.java:
```diff
@@ -0,0 +1,36 @@
+    private String author;
+    private int year;
+
+    public Book(String title, String author, int year) {
```
воспользуйся аннотациями Lombok - ```@AllArgsConstructor```, ```@ToString``` и ```@EqualsAndHashCode``` - так твой код станет более лаконичным и гибким, ведь тебе не придется менять каждый метод при появление нового поля

MarselYsup commented on src/main/java/school/faang/LibrarySystem.java:
```diff
@@ -0,0 +1,46 @@
+import java.util.HashMap;
+
+public class LibrarySystem {
+    public HashMap<Book, String> bookMap = new HashMap<>();
```
сделай final, чтобы он был неизменяемым, и никто не мог бы его поменять на другую мапу

MarselYsup commented on src/main/java/school/faang/LibrarySystem.java:
```diff
@@ -0,0 +1,46 @@
+        bookMap.put(book, location);
+    }
+
+    public void addBook(Book book, String location) {
```
в задаче не требуется делать перегруженный метод, также подмечу, что такой метод небезопасен тем, что я могу передать объект book, потом поменять любое поле этого объекта и больше возможности найти его в маппе я не смогу - можешь прочитать про работу мапы или можешь сам попробовать

MarselYsup commented on src/main/java/school/faang/LibrarySystem.java:
```diff
@@ -0,0 +1,46 @@
+    }
+
+    public void removeBook(String title, String author, int year) {
+        bookMap.remove(new Book(title, author, year));
```
```remove``` метод возвращает null, если маппа ничего не удалила, можно этим воспользоваться, чтобы информировать в консоль, что ничего не было удалено

MarselYsup commented on src/main/java/school/faang/LibrarySystem.java:
```diff
@@ -0,0 +1,46 @@
+        bookMap.remove(new Book(title, author, year));
+    }
+
+    public void removeBook(Book book) {
```
в задаче не требуется делать перегруженный метод

MarselYsup commented on src/main/java/school/faang/LibrarySystem.java:
```diff
@@ -0,0 +1,46 @@
+        }
+    }
+
+    public void findBook(Book book) {
```
в задаче не требуется делать перегруженный метод

MarselYsup commented on src/main/java/school/faang/LibrarySystem.java:
```diff
@@ -0,0 +1,46 @@
+    }
+
+    public void findBook(String title, String author, int year) {
+        for (HashMap.Entry<Book, String> book : bookMap.entrySet()) {
```
поизучай как работает мапа, тебе не нужно проходить по каждому ее элементу, достаточно просто вызвать метод у маппы ```get```, которое вернет в среднем за константное время тебе значение, вместо линейного

MarselYsup commented on src/main/java/school/faang/LibrarySystem.java:
```diff
@@ -0,0 +1,46 @@
+
+    public void printAllBooks() {
+        for (HashMap.Entry<Book, String> book : bookMap.entrySet()) {
+            System.out.println("The book: " + book.getKey() + " is located in: " + book.getValue() + ".");
```
Используй форматирование через printf, чтобы не создать большое кол-во объектов в StringPool(так как String неизменяемый и каждое складывание строк создает в памяти новую строку), а также такой подход более производительный, его проще поддерживать, особенно в случае сложных строк или когда выходные данные включают несколько типов данных

```java
System.out.printf"The book: %s is located in: %s." , book.getKey(), book.getValue());
```
можешь почитать, какие есть форматы для него под разные типы данных

MarselYsup commented on src/main/java/school/faang/LibrarySystem.java:
```diff
@@ -0,0 +1,46 @@
+
+    public void addBook(String title, String author, int year, String location) {
+        Book book = new Book(title, author, year);
+        bookMap.put(book, location);
```
можешь не создавая объект сразу класть  - 
```java
bookMap.put(new Book(title, author, year), location);
```

MarselYsup commented:
еще у тебя упала проверка PR на код стайле, вызови в gradle  checkstyleMain и посмотри где у тебя ошибки в кодстайле
![image](https://github.com/user-attachments/assets/b98bf04c-9f09-477d-a870-bc3984bbd191)

FilippovichSergey committed:
1. class Book. Added Lombok
2. class LibrarySystem. Made HashMap<Book, String> bookMap final.
3. class LibrarySystem. Removed public void addBook(Book book, String location).
4. class LibrarySystem. Updated method removeBook(String title, String author, int year).
5. class LibrarySystem. Removed public void removeBook(Book book).
6. class LibrarySystem. Removed public void findBook(Book book).
7. class LibrarySystem. Updated method findBook.
8. class LibrarySystem. Updated method printAllBooks().
9. class LibrarySystem. Updated method addBook

FilippovichSergey commented:
> еще у тебя упала проверка PR на код стайле, вызови в gradle checkstyleMain и посмотри где у тебя ошибки в кодстайле ![image](https://private-user-images.githubusercontent.com/70720056/433985450-b98bf04c-9f09-477d-a870-bc3984bbd191.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDQ3NDUzNzMsIm5iZiI6MTc0NDc0NTA3MywicGF0aCI6Ii83MDcyMDA1Ni80MzM5ODU0NTAtYjk4YmYwNGMtOWYwOS00NzdkLWE4NzAtYmMzOTg0YmJkMTkxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA0MTUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNDE1VDE5MjQzM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTA4YzJkOTM3NjAwM2UyNjIzYTJiMjI5MTMzOTQ1NDZhYTkyZjQ1OGU1NjYzOGEwOTEzNTYzYWU0YzZhMjY5NDAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.R7dlAFwZgBWzcqg2193C2QhyR0nI4ZgG2o9IUrywkDE)

подскажи, пожалуйста, где найти checkstyleMain?
я смог найти только \config\checkstyle\checkstyle.xml там у меня горит красным http://www.puppycrawl.com/dtds/configuration_1_3.dtd

MarselYsup commented on src/main/java/school/faang/LibrarySystem.java:
```diff
@@ -0,0 +1,27 @@
+
+    public String findBook(String title, String author, int year) {
+        Book book = new Book(title, author, year);
+        return bookMap.getOrDefault(book, "The book is not found.");
```
На практике, если ты возвращаешь результат, а не null, то это воспринимается как успешный вызов метода, поэтому лучше возвращать null
или просто вызывать метод get
```java
bookMap.get(book)
```
так в случае отсутствия книги, вернется null

MarselYsup replied:
не поправил коммент

FilippovichSergey replied:
исправил метод.
У меня появились 2 ошибки в классе LibrarySystem, но при этом код запускается

MarselYsup commented on src/main/java/school/faang/Book.java:
```diff
@@ -0,0 +1,37 @@
+import java.util.Objects;
+import lombok.Data;
+@Data
+public class Book {
```
Добавь @AllArgsConstructor @ToString и @EqualsAndHashCode, а реализованные методы удали, Lombok сам будет генерить

MarselYsup commented:
![image](https://github.com/user-attachments/assets/b11825ef-253e-4637-b0f6-7405c77d2d28)
Вообще если пользуешься Idea от JetBrains то справа есть вкладка gradle
вот файлик тебе который он генерит по твоим ошибкам
[main.pdf](https://github.com/user-attachments/files/19764942/main.pdf)

FilippovichSergey committed:
1. class Book. Added annotations @AllArgsConstructor, @ToString, @EqualsAndHashCode
2. class Book. Removed unused methods
3. class LibrarySystem. Remove space

FilippovichSergey committed:
class LibrarySystem. Updated method findBook

FilippovichSergey committed:
updated style in class Main
